services/vrf/vrf.go

```diff
@@ -0,0 +1,323 @@
+var P = bigFromBase10(
+	"21888242871839275222246405745257275088696311157297823662689037894645226208583")
+
+// Order is the number of rational points on the curve in GF(P) (group size)
+var Order = bigFromBase10(
+	"21888242871839275222246405745257275088548364400416034343698204186575808495617")
+
+// Compensate for awkward big.Int API.
+var bi = big.NewInt
+var zero, one, two, three, four = bi(0), bi(1), bi(2), bi(3), bi(4)
+
+func i() *big.Int                                    { return new(big.Int) }
+func add(addend1, addend2 *big.Int) *big.Int         { return i().Add(addend1, addend2) }
+func div(dividend, divisor *big.Int) *big.Int        { return i().Div(dividend, divisor) }
+func equal(left, right *big.Int) bool                { return left.Cmp(right) == 0 }
+func exp(base, exponent, modulus *big.Int) *big.Int  { return i().Exp(base, exponent, modulus) }
+func lsh(num *big.Int, bits uint) *big.Int           { return i().Lsh(num, bits) }
+func mul(multiplicand, multiplier *big.Int) *big.Int { return i().Mul(multiplicand, multiplier) }
+func mod(dividend, divisor *big.Int) *big.Int        { return i().Mod(dividend, divisor) }
+func sub(minuend, subtrahend *big.Int) *big.Int      { return i().Sub(minuend, subtrahend) }
+
+// Sadly, the hidden interfaces in bn256 require us to duplicate some
+// functionality.
+
+var (
+	// (P-1)/2: Half Fermat's Little Theorem exponent
+	eulersCriterionPower = div(sub(P, one), two)
+	// (P+1)/4: As long as P%4==3 and n=x^2 in GF(P), n^((P+1)/4)=±x
+	sqrtPower = div(add(P, one), four)
+)
+
+// IsSquare returns true iff x = y^2 for some y in GF(p)
+func IsSquare(x *big.Int) bool {
+	return equal(one, exp(x, eulersCriterionPower, P))
+}
+
+// SquareRoot returns a s.t. a^2=x. Assumes x is a square
+func SquareRoot(x *big.Int) *big.Int {
+	return exp(x, sqrtPower, P)
+}
+
+// YSquared returns x^3+3 mod P
+func YSquared(x *big.Int) *big.Int {
+	return mod(add(exp(x, three, P), three), P)
+}
+
+// IsCurveXOrdinate returns true iff there is y s.t. y^2=x^3+3
+func IsCurveXOrdinate(x *big.Int) bool {
+	return IsSquare(YSquared(x))
+}
+
+func packUint256s(xs ...*big.Int) ([]byte, error) {
+	mem := bytes.Buffer{}
+	for _, x := range xs {
+		word, err := utils.EVMWordBigInt(x)
+		if err != nil {
+			return []byte{}, err
+		}
+		n, err := mem.Write(word)
+		if n != 32 {
+			return []byte{}, fmt.Errorf(
+				"Failed to write as uint256: %v", x)
+		}
+		if err != nil {
+			return []byte{}, err
+		}
+	}
+	if mem.Len() != len(xs)*32 {
+		panic(fmt.Errorf("Package of %v uint256s unexpected length, %v",
+			len(xs), mem.Len()))
+	}
+	return mem.Bytes(), nil
+}
+
+// NewCurvePoint returns the bn256.G1 point corresponding to (x, y)
+func NewCurvePoint(x, y *big.Int) (*bn256.G1, error) {
+	p := new(bn256.G1)
+	packed, err := packUint256s(x, y)
+	if err != nil {
+		return &bn256.G1{}, err
+	}
+	b, err := p.Unmarshal(packed) // Unmarshal's comment lies about return type
+	if len(b) != 0 {
+		panic(fmt.Errorf(
```

Comment:
Stay calm, no need to panic. Preferably we'd just return an error.

If you were relying on it to ensure Generator is properly initialized, we could wrap the call to NewCurvePoint in a private function just for initialization.